File: checker.py
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from colorama import init, Fore
import time

# Inisialisasi colorama
init(autoreset=True)

def check_proxy(proxy, retries=3):
    proxies = {
        "http": proxy,
        "https": proxy,
    }
    for attempt in range(retries):
        try:
            # Mengirim permintaan GET ke sebuah URL
            response = requests.get("http://httpbin.org/ip", proxies=proxies, timeout=2)  # Timeout 2 detik
            if response.status_code == 200:
                return proxy, True
        except requests.exceptions.RequestException:
            pass
    return proxy, False

def write_result_to_file(proxy, is_active):
    status = "active" if is_active else "inactive"
    with open("active_proxies.txt" if is_active else "inactive_proxies.txt", 'a') as file:
        file.write(proxy + "\n")

def check_proxies_from_file(file_path):
    with open(file_path, 'r') as file:
        proxies = [proxy.strip() for proxy in file.readlines() if proxy.strip()]  # Menghapus whitespace dan baris kosong

    # Meningkatkan jumlah thread
    with ThreadPoolExecutor(max_workers=20) as executor:  # Menggunakan 20 thread
        future_to_proxy = {executor.submit(check_proxy, proxy): proxy for proxy in proxies}

        for future in as_completed(future_to_proxy):
            proxy, is_active = future.result()
            status = "active" if is_active else "inactive"
            
            # Menampilkan hasil di konsol dengan warna
            if is_active:
                print(f"{proxy} {Fore.GREEN}{status}{Fore.RESET}")  # Warna hijau untuk active
            else:
                print(f"{proxy} {Fore.RED}{status}{Fore.RESET}")  # Warna merah untuk inactive
            
            # Menyimpan hasil ke file secara real-time
            write_result_to_file(proxy, is_active)

if __name__ == "__main__":
    # Ganti dengan path ke file yang berisi daftar proxy
    file_path = "proxy.txt"  # Contoh: "proxies.txt"
    
    start_time = time.time()  # Mulai pengukuran waktu
    check_proxies_from_file(file_path)
    end_time = time.time()  # Akhiri pengukuran waktu

    print(f"\nHasil proxy aktif disimpan di 'active_proxies.txt'")
    print(f"Hasil proxy tidak aktif disimpan di 'inactive_proxies.txt'")
    print(f"Waktu eksekusi: {end_time - start_time:.2f} detik")

File: checker.js
import fs from 'fs';
import axios from 'axios';
import chalk from 'chalk';

/**
 * Fungsi untuk memeriksa apakah proxy aktif
 * @param {string} proxyUrl - URL proxy dalam format 'http://username:password@proxyserver:port'
 * @returns {Promise<{url: string, status: string}>} - Mengembalikan objek dengan URL dan status
 */
async function checkProxy(proxyUrl) {
    try {
        const response = await axios.get('http://httpbin.org/ip', {
            proxy: {
                host: proxyUrl.split(':')[2].split('@')[1], // Mengambil host dari URL proxy
                port: parseInt(proxyUrl.split(':')[2].split('@')[0]), // Mengambil port dari URL proxy
            },
            timeout: 5000 // Timeout 5 detik
        });

        return { url: proxyUrl, status: 'active' }; // Proxy aktif
    } catch (error) {
        return { url: proxyUrl, status: 'inactive' }; // Proxy tidak aktif
    }
}

/**
 * Fungsi untuk membaca file dan memeriksa semua proxy
 * @param {string} filePath - Path ke file yang berisi daftar proxy
 */
async function checkProxiesFromFile(filePath) {
    try {
        const data = fs.readFileSync(filePath, 'utf8');
        const proxies = data.split('\n').filter(line => line.trim() !== ''); // Menghapus baris kosong

        // Buat atau bersihkan file output
        fs.writeFileSync('active_proxies.txt', '', 'utf8');
        fs.writeFileSync('inactive_proxies.txt', '', 'utf8');

        for (const proxy of proxies) {
            const result = await checkProxy(proxy.trim());
            const output = `${result.url} ${result.status}`;

            // Menyimpan hasil ke file secara real-time
            if (result.status === 'active') {
                console.log(chalk.green(output)); // Tampilkan aktif dengan warna hijau
                fs.appendFileSync('active_proxies.txt', result.url + '\n', 'utf8');
            } else {
                console.log(chalk.red(output)); // Tampilkan tidak aktif dengan warna merah
                fs.appendFileSync('inactive_proxies.txt', result.url + '\n', 'utf8');
            }
        }

        console.log('Hasil pemeriksaan telah disimpan ke dalam file active_proxies.txt dan inactive_proxies.txt');

    } catch (error) {
        console.error('Error membaca file:', error.message);
    }
}

// Contoh penggunaan
const filePath = 'proxies.txt'; // Ganti dengan path ke file proxy Anda
checkProxiesFromFile(filePath);
